File: my-solutions/0146-lru-cache/solution.py
class Node:
    def __init__(self, key, val):
        self.key, self.val = key, val
        self.left = self.right = None

class LRUCache:

    def __init__(self, capacity: int):
        self.right, self.left = Node(0, 0), Node(0, 0)
        self.right.next, self.left.prev = None, None
        self.right.prev, self.left.next = self.left, self.right
        self.cap = capacity
        self.cache = {}
    
    def remove(self, node):
        prev, nxt = node.prev, node.next
        prev.next, nxt.prev = nxt, prev

    def insert(self, node): # insert at right
        prev, nxt = self.right.prev, self.right
        prev.next = nxt.prev = node
        node.prev, node.next = prev, nxt

    def get(self, key: int) -> int:
        if key in self.cache:
            self.remove(self.cache[key])
            self.insert(self.cache[key])
            return self.cache[key].val
        else:
            return -1
        
    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.cache[key].val = value
            self.remove(self.cache[key])
            self.insert(self.cache[key])
        else:
            if len(self.cache) == self.cap:
                lru = self.left.next
                self.remove(lru)
                del self.cache[lru.key]

                

            node = Node(key, value)
            self.cache[key] = node
            self.insert(node)
            
            # temp = self.left
            # while(temp):
            #     print(temp.key, temp.val)
            #     temp = temp.next
            # print("")
        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: my-solutions/0146-lru-cache/solution.cpp
class LRUCache {
    int cap;
    std::unordered_map<int, std::list<std::pair<int, int>>::iterator> cacheMap;
    std::list<std::pair<int, int>> cacheList;

public:
    LRUCache(int capacity) : cap(capacity) {}

    int get(int key) {
        auto mapIt = cacheMap.find(key);
        if (mapIt == cacheMap.end()) {
            return -1;
        } else {
            auto listIt = mapIt->second; // Get the list iterator from the map
            int value = listIt->second;  // Extract the value from the pair
            cacheList.splice(cacheList.begin(), cacheList, listIt); // Move to front
            return value;
        }
    }

    void put(int key, int value) {
        auto it = cacheMap.find(key);
        if (it == cacheMap.end()) {
            // If cache is full, evict the least recently used item
            if (cacheList.size() == cap) {
                int lruKey = cacheList.back().first; // Get key of LRU item
                cacheMap.erase(lruKey);              // Remove from map
                cacheList.pop_back();                // Remove from list
            }
            // Insert new key-value pair
            cacheList.push_front({key, value});
            cacheMap[key] = cacheList.begin();
        } else {
            // Update existing key
            auto listIt = it->second;         // Get the list iterator
            listIt->second = value;           // Update the value
            cacheList.splice(cacheList.begin(), cacheList, listIt); // Move to front
        }
    }
};
